pkg/edgeview/src/copyfile.go

```diff
@@ -146,36 +171,51 @@ func runCopy(opt string) {
 	}()
 
 	<-ahead
+	if tarDirSize != nil { // the caller will handle the file transfer operation
+		log.Functionf("runCopy: received from client to go ahead")
+		return nil
+	}
+
 	if !readerRunning {
-		return
+		return nil
 	}
 	f, err := os.Open(file)
 	if err != nil {
 		fmt.Printf("os open error %v\n", err)
-		return
+		return fmt.Errorf("cp command error")
 	}
 	defer f.Close()
 
-	buffer := make([]byte, 8192)
+	// the delay is only related to the transfer chunksize, independent of total size of the file
+	sleepDuration := getSleepDuration(chunkSize)
+	buffer := make([]byte, chunkSize)
 	totalBytes := 0
 	for {
+		if lostClientPeer { // in the middle of transferring file, if client is gone, jump out
+			log.Functionf("createArchive: client is gone. Stop transferring files")
+			break
+		}
 		n, err := f.Read(buffer)
 		if err != nil {
 			fmt.Printf("file read error %v\n", err)
-			return
+			return fmt.Errorf("cp command error")
 		}
 
 		err = addEnvelopeAndWriteWss(buffer[:n], false)
 		if err != nil {
 			fmt.Printf("file write to wss error %v\n", err)
-			return
+			return fmt.Errorf("cp command error")
 		}
 		totalBytes += n
 		if totalBytes >= int(cfile.Size) {
 			break
 		}
+		if !disableRateLimit {
+			time.Sleep(sleepDuration)
```

Comment:
Shouldn't this be shortened by the time that was spent sending the chunk?
Right now I do not see how the actual speed of the link is taken into the account (could be already 5Mbps or less).

Reply:
Also, should we also consider the actual number of bytes received (`n` vs. `chunkSize`)?

Reply:
@milan-zededa try to be simple here, I give 20% extra for this ratelimit here. Not meant to be very accurate, but don't want user to blast the traffic towards the controller side. We pick an artificial target rate 5Mbps for that, i think it's ok if its 5.5 or 4.8.
yes, I could have used 'n' vs chunkSize, but this is copying a single file, only the last chunk would be smaller than the chunkSize here, so this does not matter too much. In the 'tar' op case, we do use the actually 'n', since each file can be very small.

Reply:
OK, understood